cpp/include/algorithms.hpp

```diff
@@ -662,24 +664,26 @@ void louvain(GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
  *                                   Supported value : int (signed, 32-bit)
  * @tparam weight_t                  Type of edge weights. Supported values : float or double.
  *
+ * @param[in]  handle                Library handle (RAFT). If a communicator is set in the handle,
  * @param[in]  graph                 input graph object (CSR)
- * @param[out] final_modularity      modularity of the returned clustering
- * @param[out] num_level             number of levels of the returned clustering
  * @param[out] clustering            Pointer to device array where the clustering should be stored
  * @param[in]  max_iter              (optional) maximum number of iterations to run (default 100)
  * @param[in]  resolution            (optional) The value of the resolution parameter to use.
  *                                   Called gamma in the modularity formula, this changes the size
  *                                   of the communities.  Higher resolutions lead to more smaller
  *                                   communities, lower resolutions lead to fewer larger
  * communities. (default 1)
+ *
+ * @return                           a pair containing:
+ *                                     1) number of levels of the returned clustering
+ *                                     2) modularity of the returned clustering
  */
 template <typename vertex_t, typename edge_t, typename weight_t>
-void leiden(GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
-            weight_t &final_modularity,
-            int &num_level,
-            vertex_t *leiden_parts,
-            int max_iter        = 100,
-            weight_t resolution = weight_t{1});
+std::pair<int, weight_t> leiden(raft::handle_t const &handle,
```

Comment:
I think we'd better use size_t for number of levels and maximum iterations (this will have negligible impact in memory footprint and performance while more future ready).

std::pair<size_t, weight_t> leiden(... , size_t max_iter = 100, ...);

Reply:
I could make that change.  These parameters represent the number of iterations of the outer loop of Louvain, each loop shrinks the graph by a significant amount (don't have a hard heuristic on that).  I suspect we'd need a graph with something on the order of 2^256 edges in order to exceed 2^32 Louvain levels... but that's intuition not based on any measurements.

Reply:
I don't think this will ever lead to overflow here, either.

My preference on size_t over int is that 1) size_t tells more about the nature of the variable than int and 2) consistency (it may not lead to overflow here but can happen elsewhere).

I generally try to avoid using (unsigned) short, (unsigned) int, (unsigned) long unless this is used by the external library I am interfacing with (e.g. MPI uses int, so I use int for number of processes...).

But this is just my personal preference, and just a recommendation.

Reply:
Done